### PR TITLE
chore (keyv): remove as types now shipped

### PR DESCRIPTION
**Pull request by 43081j**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).
- [x] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [x] Delete the package's directory.
- [x] Add it to `notNeededPackages.json`.

keyv ships its own types here:
https://github.com/jaredwray/keyv/blob/main/packages/keyv/src/index.d.ts

i specified "since 4.5.0" just because its been this way for a while and im not sure which exact version introduced it

this also applies to all the packages in the keyv org. 

I updated a couple of referencing packages too because they pulled keyv in but probably didn't need to

### CODE REVIEW

**Comment by 43081j** [comment]
Will fix ci soon, my bad 👍

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## got ([<kbd>unpkg</kbd>](https://unpkg.com/browse/got@latest/))
was missing the following properties:
1. create
1. mergeInstances
1. defaults
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 8e6615b4dab6f524c6e95cc09c58e2a101eaf312
</p>

**Comment by andrewbranch** [comment]
> i specified "since 4.5.0" just because its been this way for a while and im not sure which exact version introduced it

unpkg makes it pretty easy to find out. It was 4.2.0:

https://unpkg.com/browse/keyv@4.2.0/src/

**Comment by 43081j** [comment]
nice, didn't think of that. have updated the versions now

**Review by andrewbranch — APPROVED** [review]

**Comment by 43081j** [comment]
ready to merge

**Comment by Ansile** [comment]
Looks like this causes some issues - https://github.com/DefinitelyTyped/DefinitelyTyped/issues/62793

Could this be reverted/fixed by empty .d.ts file, maybe?

**Comment by Ansile** [comment]
@43081j what was the reason for removing index.d.ts file altogether and not leaving it empty? Did this cause some sort of types collision / shadowing?

**Comment by 43081j** [comment]
@Ansile i don't control the versioning when removing a package, the DT tooling decides that so afraid i can't answer that

may be an issue we need to raise in the dt tools repo.

i suspect the issue you've referenced is a red herring. i think its more likely somewhere in your dependency tree, you have types which depend on keyv. those are expecting you to have `@types/keyv` available (which you no longer do), so they try find the types in keyv itself and fail because its an older version which didn't ship types.

it'd be helpful if we could know what that dependency is. somewhere it sounds like you have a dependency which depends on keyv.

**Comment by Ansile** [comment]
> @Ansile i don't control the versioning when removing a package, the DT tooling decides that so afraid i can't answer that
> 
> may be an issue we need to raise in the dt tools repo.

Hm, i did create the issue - https://github.com/DefinitelyTyped/DefinitelyTyped/issues/62793

Maybe it's a bit of an unhandled case

> i suspect the issue you've referenced is a red herring. i think its more likely somewhere in your dependency tree, you have types which depend on keyv. those are expecting you to have `@types/keyv` available (which you no longer do), so they try find the types in keyv itself and fail because its an older version which didn't ship types.

I do have `@types/keyv` installed though. 

image

The case behind the error is exactly it **being** installed **but not having index.d.ts file**

> it'd be helpful if we could know what that dependency is. somewhere it sounds like you have a dependency which depends on keyv.

That would be `@types/cacheable-request@6.0.2`

image

It looks like it's also been removed from definitelyTyped, but it's previous versions depends on "@types/keyv": "*", which actually means that major version would not have helped. That's pretty fragile :(

> also, importantly, if you don't choose to upgrade @types/keyv, everything works fine at least right? we haven't broken existing code, just case where someone upgrades. sure there's something to solve here still but just want to confirm that

That is the problem i'm trying to address. I did not update/upgrade anything willingly. Any fresh install previously was working, now it's not.
Of course if i force the install of a previous version, it works. That is indeed a workaround, but i'm afraid many other developers in many packages would have to do it.

**Comment by 43081j** [comment]
i more meant an issue that when we remove packages, we should bump the major version possibly. though there may be reasons we don't do that im just not aware of.

i think i understand now.

you're getting problems because you've updated to the latest `@types/keyv` (since its under a patch version). that is a stub which doesn't really contain anything. 

any code you then directly have which uses keyv will fail until you also upgrade keyv itself (so you pull in the new types).

similarly, any dependencies you have which depend on `@types/keyv` with a loose enough semver range will also pull in the stub. so they will no longer work unless they themselves upgrade the keyv they depend on too.

ultimately this drills down to the fact that it shipped in a patch version i think.

i don't think anything is really broken here. anyone who wants to upgrade should remove `@types/keyv` and upgrade `keyv`. i think its questionable though that this shipped in a patch version.

**Comment by Ansile** [comment]
@43081j yes, i guess it boils down to:
- this landing in patch instead of major. If you can add some input about it as the change author in the issue https://github.com/DefinitelyTyped/DefinitelyTyped/issues/62793, that would be great. My specific problem can be fixed by a workaround on my part, but i would like to prevent something like this from happening again.

- @types/cacheable-request having `@types/keyv: "*"` in the dependencies in the first place. I don't know if it's an another part of DefinitelyTyped principles. If so, it seems really misguided in nature, as there is really no way to guarantee forward compatibility for an infinitely long time, and semver is supposed to handle exactly that.
Basically, IMO, any "*" in dependencies is a problem waiting to happen.

**Comment by jaredwray** [comment]
@Ansile - lets go ahead and revert this as it is breaking many older versions.

**Comment by 43081j** [comment]
@jaredwray further explanation would be great.

the old version is available on NPM as expected, anyone who chooses to upgrade can use the shipped types.

what makes you believe this needs reverting?

**Comment by jaredwray** [comment]
It looks like we will need to also add in support for 4.5 and keep it active until version 5
________________________________

**Comment by 43081j** [comment]
for what reason?

4.5 ships types:
https://unpkg.com/browse/keyv@4.5.0/src/index.d.ts

**Comment by 43081j** [comment]
are the shipped types not correct?

just trying to understand as reverting this shouldn't be necessary except in one case: that the pre-shipped-types version had incomplete types here we now want to correct
